### PR TITLE
Adds fields supporting no-sandbox mode to IR module.

### PR DESCRIPTION
**Pull request by matthias-blume**
Adds fields supporting no-sandbox mode to IR module, as well as the code to populate them.

### CODE REVIEW

**Comment by matthias-blume** [comment]
So I assume we have to turn off the CI / build (windowns-latest) check.  How is this done?

**Comment by dimitry-** [comment]
can you add a test for this?

**Comment by matthias-blume** [comment]
I believe this is ready to go.

File: src/binary-reader-ir.cc [hunk 1]
```diff
@@ -1552,6 +1616,42 @@ Result BinaryReaderIR::OnNameEntry(NameSectionSubsection type,
...
+    case RelocType::MemoryAddressSLEB64:
+    case RelocType::MemoryAddressLEB:
+    case RelocType::MemoryAddressLEB64:
+      module_->memory_address_operand_offsets_.insert(offset);
+      break;
+    case RelocType::MemoryAddressI32:
+      module_->memory_address_32_data_initializer_offsets_.insert(offset);
+      break;
+    case RelocType::MemoryAddressI64:
+      module_->memory_address_64_data_initializer_offsets_.insert(offset);
```
Comment (sbc100) [inline src/binary-reader-ir.cc]: Wouldn't you also need to store both the index and the offset to do something useful with this later?

Should we store all the relocation information in the IR?  e.g `module->data_relocs_.push_back(...)` and `module->code_relocs_.push_back(...)`  ?

Comment (matthias-blume) [reply src/binary-reader-ir.cc]: The index in question is the memory- or table-index?  In no-sandbox mode those are assumed to be 0.  (This is already explicitly checked elsewhere.)  If the index refers to something else, please let me know.  There are definitely still big gaps in my understanding...

Maybe there is a case to be made to comprehensively store all the reloc info in the IR.  For now I was just going for a minimalist design that just supports no-sandbox mode and nothing more.  (In particular, I've chosen ordered sets and maps where I know that those will come in handy for what I want to do.)

Maybe this can be resolved independently and at a later point?

Comment (sbc100) [reply src/binary-reader-ir.cc]: The indexs here are things like function indexes yes and segment indexes.    Without the information I can't see how you can apply the relocations (whether that is somehow statically in the generated C code, or during startup).  

Somebody has to apply the relocations, and I don't see how you can do that if the only information you have is the offset at which the relocatation lives.

For example, a relocation might be "index: printf, offset: 10"  which would mean that the address of printf is to be placed at offset 10 withing the segment.    If you only stash the offset 10 you wouldn't know what to put there.   
With data symbols you also need the addend.

In other words, I think you will find that you need all of the information about each relocation in order to apply/resolve them.

Comment (matthias-blume) [reply src/binary-reader-ir.cc]: I was able to make this work without storing the indices.  But I see now that what I was trying to do might actually be clumsy and can be done better.

Basically, the way it was supposed to work is this:  I'm looking at some ixx.const i instruction.  I have its offset and its value of i in hand.  I consult the set of offsets that I got from relocations and ask "Is my offset in that set?".  If not, I write the value of i, since it is just a plain constant.  If it is in the set of function relocation offsets, I know that i refers to a table entry.  So I use i as an index into the table->function index mapping (which is also being prepared).

Correct me if I'm wrong, but it now looks like I can skip the second step, because the function index that I need is also stored as part of the relocations, so I can construct a direct mapping and can avoid the second lookup.

That would indeed be good, because preparing the mapping from table indices to corresponding function indices (and also the one from offsets to memory segments) is fraught with significant trouble, because processing element- and data segments relies on them having a specific format (in particular: their offset being given as a single ixx.const instruction).

In summary:

- For relocations that point at function pointers the index is the function index (i.e., points at the corresponding entry in module_->funcs).
- For relocations that point at memory addresses the index is the data segment index (i.e., points at the corresponding entry in module_->data_segments)

Is that correct?  If so, it will definitely make my life a lot easier! :)

Comment (sbc100) [reply src/binary-reader-ir.cc]: Yes, exactly.   There may be times you can infer the index by looking at the location of the relocations.  e.g. we try to have i32.const have a meaningful value, even when it has a relocation.

However, IIRC there are cases where we can't give the meaningful value, and you should certainly be using the relocation entry as the source of truth, not the existing content that the location.

Comment (matthias-blume) [reply src/binary-reader-ir.cc]: (Non-)Progress report:  My initial experiments seem to indicate that at least for function pointer relocations (ixx.const instructions that represent taking the address of a function) the index value in the relocation does not reference the correct entry in module_->funcs.  I am trying to read the linking spec, but I am having trouble making sense of it.  In any case, I'm seeing index values pointing to the wrong functions, and other index values that aren't even within the range of module_->funcs.

I'm clearly still missing something.  I'll bring it up in tomorrow's meeting.

Comment (sbc100) [reply src/binary-reader-ir.cc]: It could be that you are running into one of the slightly confusing thing about the global, and function, index spaces:  They start with the imported globals/functions and go on to the local global/functions.   So, in a module with no imports the function index space matches the index into the defined functions.    But, if you have imported functions they always occupy the start of the index space.

Comment (matthias-blume) [reply src/binary-reader-ir.cc]: That checks out ... to some extent.  For reloc index values that fall within the appropriate range of indices into module_->funcs it is indeed the case that adjusting by module_->num_func_imports fixes the problem.  However, I am also seeing an index value that is strictly outside that range.  In fact, it is equal to module_->funcs.size().
The relocation in question corresponds to a function pointer taken of an imported function.

With my old method of using the actual constant in the instruction and indirecting through the contents of the function table I was able to obtain the correct entry of module_->funcs for both local and imported functions.  But now it looks like something else is required for imports, and I don't know what that is.

Comment (matthias-blume) [reply src/binary-reader-ir.cc]: I got this to work using symbol table information.

**Comment by matthias-blume** [comment]
This looks ready to go now.  Checks all pass.

**Review by dimitry- — APPROVED** [review]
